README.md
description: A customizable Flutter library for horizontal-scrolling picker tabs.

## Features

<img src="https://raw.githubusercontent.com/aigongzuo/horizontal_picker_tabs/main/img_0.webp" width="240px" height="426px"/>
<img src="https://raw.githubusercontent.com/aigongzuo/horizontal_picker_tabs/main/img_1.webp" width="426px" height="240px"/>

## Getting started

dart pub add horizontal_picker_tabs

## Usage

```dart
Scaffold(
resizeToAvoidBottomInset: false,
backgroundColor: Colors.black,
body: Stack(children: [
PickerTabBarView(
controller: _tabController,
children: getTabViewList(),
),

//装饰底部卡尺
Positioned(
bottom: 0,
left: 0,
right: 0,
child: Container(
height: 100,
color: Colors.black54,
),
),

//装饰底部卡尺
const Positioned(
bottom: 14,
left: 0,
right: 0,
child: Image(
height: 20,
fit: BoxFit.fill,
image: AssetImage("assets/images/lm_horizontal_picker_bottom_icon.png"),
),
),

Positioned(
bottom: 14,
left: 0,
right: 0,
child: Center(
child: Container(
margin: const EdgeInsets.only(top: 2),
width: 2,
height: 16,
color: const Color(0xFF7F5EFF),
),
),
),

Column(
mainAxisAlignment: MainAxisAlignment.end,
children: [
HorizontalPickerTabBar(
controller: _tabController,
padding: const EdgeInsets.only(top: 15, bottom: 5),
// indicatorPadding: const EdgeInsets.all(2),
labelColor: Colors.white,
labelStyle: labelStyle,
unselectedLabelColor: const Color(0x99FFFFFF),
unselectedLabelStyle: unselectedLabelStyle,
tabs: widget.tabs,
),
],
)
]),
);
```



CHANGELOG.md
## 1.0.0
* 1.0.0 release publish.

## 1.0.1
* fix homepage URL


test/horizontal_picker_tabs_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:horizontal_picker_tabs/horizontal_picker_tabs.dart';

void main() {
  test('adds one to input values', () {
    // final calculator = Calculator();
    // expect(calculator.addOne(2), 3);
    // expect(calculator.addOne(-7), -6);
    // expect(calculator.addOne(0), 1);
  });
}


lib/horizontal_picker_tabs.dart
import 'package:flutter/material.dart';
import 'package:horizontal_picker_tabs/picker_tabs.dart';

class HorizontalPickerTabBar extends StatefulWidget {
  final List<String> tabs;
  final TabController? controller;
  final EdgeInsetsGeometry? padding;
  final Color? labelColor;
  final Color? unselectedLabelColor;
  final TextStyle labelStyle;
  final TextStyle unselectedLabelStyle;

  const HorizontalPickerTabBar({
    super.key,
    required this.tabs,
    this.controller,
    this.padding,
    this.labelColor,
    this.unselectedLabelColor,
    required this.labelStyle,
    required this.unselectedLabelStyle,
  });

  @override
  State<StatefulWidget> createState() {
    return HorizontalPickerTabBarState();
  }
}

class HorizontalPickerTabBarState extends State<HorizontalPickerTabBar> {
  double marginTabBarStart = 0;
  double marginTabBarEnd = 0;
  double tabPaddingLeftRight = 20;

  @override
  Widget build(BuildContext context) {
    return PickerTabBar(
      controller: widget.controller,
      padding: widget.padding,
      // indicatorPadding: const EdgeInsets.all(2),
      indicatorPadding: EdgeInsets.zero,
      // 设置为零以去掉间距
      indicator: const BoxDecoration(color: Colors.transparent),
      // indicatorSize: TabBarIndicatorSize.label,
      labelPadding: EdgeInsets.zero,
      labelColor: widget.labelColor,
      labelStyle: widget.labelStyle,
      unselectedLabelColor: const Color(0x99FFFFFF),
      unselectedLabelStyle: widget.unselectedLabelStyle,
      isScrollable: true,
      tabs: getTitleViewList(widget.tabs),
    );
  }

  getTitleViewList(List<String> tabs) {
    double screenWidth = MediaQuery.of(context).size.width;
    marginTabBarStart = screenWidth / 2 -
        getTextWidth(tabs[0], widget.labelStyle) / 2 -
        tabPaddingLeftRight;
    marginTabBarEnd = screenWidth / 2 -
        getTextWidth(tabs[tabs.length - 1], widget.labelStyle) / 2 -
        tabPaddingLeftRight;

    //遍历dataListList
    List<Widget> list = [];
    int len = tabs.length;
    for (int i = 0; i < len; i++) {
      list.add(Container(
        padding: EdgeInsetsDirectional.only(
            start: tabPaddingLeftRight, end: tabPaddingLeftRight),
        margin: EdgeInsetsDirectional.only(
          start: i == 0 ? marginTabBarStart : 0,
          end: i == len - 1 ? marginTabBarEnd : 0,
        ),
        child: Column(
          children: [
            Text(tabs[i]),
            const SizedBox(
              height: 34,
            ),
          ],
        ),
      ));
    }
    return list;
  }

  isRtl() {
    TextDirection currentTextDirection = getTextDirection();
    // 判断是否是从右向左的文本布局
    return currentTextDirection == TextDirection.rtl;
  }

  // 获取当前文本的布局方向
  TextDirection getTextDirection() {
    return Directionality.of(context);
  }

  double getTextWidth(String? text, TextStyle style) {
    final textPainter = TextPainter(
      text: TextSpan(
        text: text ?? "",
        style: style, // You can set your desired text style
      ),
      textDirection: isRtl() ? TextDirection.rtl : TextDirection.ltr,
    );
    textPainter.layout();
    return textPainter.width; // This is important to calculate the text width
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:horizontal_picker_tabs/horizontal_picker_tabs.dart';
import 'package:horizontal_picker_tabs/picker_tabs.dart';

void main() {
  runApp(MaterialApp(home: MyApp()));
}

class MyApp extends StatefulWidget {
  List<String> tabs = ["tabs1", "tabs2", "tabs3", "tabs4", "tabs5"];

  MyApp({Key? key}) : super(key: key);

  @override
  State<StatefulWidget> createState() {
    return MyAppState();
  }
}

class MyAppState extends State<MyApp> with SingleTickerProviderStateMixin {
  late TabController _tabController;

  int currentPage = 0;
  TextStyle labelStyle =
      const TextStyle(fontSize: 16, fontWeight: FontWeight.w600);
  TextStyle unselectedLabelStyle = const TextStyle(
    fontSize: 16,
    fontWeight: FontWeight.w600,
  );

  @override
  void initState() {
    super.initState();
    _tabController = TabController(vsync: this, length: widget.tabs.length)
      ..addListener(() {
        if (currentPage != _tabController.index) {
          setState(() {
            currentPage = _tabController.index;
          });
        }
      });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      resizeToAvoidBottomInset: false,
      backgroundColor: Colors.black,
      body: Stack(children: [
        PickerTabBarView(
          controller: _tabController,
          children: getTabViewList(),
        ),

        //装饰底部卡尺
        Positioned(
          bottom: 0,
          left: 0,
          right: 0,
          child: Container(
            height: 100,
            color: Colors.black54,
          ),
        ),

        //装饰底部卡尺
        const Positioned(
          bottom: 14,
          left: 0,
          right: 0,
          child: Image(
            height: 20,
            fit: BoxFit.fill,
            image: AssetImage(
                "assets/images/lm_horizontal_picker_bottom_icon.png"),
          ),
        ),

        Positioned(
          bottom: 14,
          left: 0,
          right: 0,
          child: Center(
            child: Container(
              margin: const EdgeInsets.only(top: 2),
              width: 2,
              height: 16,
              color: const Color(0xFF7F5EFF),
            ),
          ),
        ),

        Column(
          mainAxisAlignment: MainAxisAlignment.end,
          children: [
            HorizontalPickerTabBar(
              controller: _tabController,
              padding: const EdgeInsets.only(top: 15, bottom: 5),
              // indicatorPadding: const EdgeInsets.all(2),
              labelColor: Colors.white,
              labelStyle: labelStyle,
              unselectedLabelColor: const Color(0x99FFFFFF),
              unselectedLabelStyle: unselectedLabelStyle,
              tabs: widget.tabs,
            ),
          ],
        )
      ]),
    );
  }

  @override
  void dispose() {
    //释放 tabController
    _tabController.dispose();
    super.dispose();
  }

  getTabViewList() {
    List<Widget> list = [];
    for (int i = 0; i < widget.tabs.length; i++) {
      list.add(Image.asset("assets/images/img_$i.webp", fit: BoxFit.cover));
    }
    return list;
  }
}
